File: remove_dups_from_linkedlist/remove_dups_from_linkedlist.cpp
/* 
PROMPT: Write code to remove duplicates from an unsorted linked list.
*/

#include <iostream>
#include <set>
using namespace std;

struct node_t {
        const node_t * next;
        int value;
        node_t(int v, const node_t * n): next(n), value(v) {};
        node_t(int v): value(v), next(nullptr) {};
};


void remove_duplicates(node_t *& head){
    /* General strategy:
        - Create a set of values seen
        - Traverse the linked list and for each node:
            - If it's value is already in the set
                -Remove that node from the linked list
            - Else: add it's value to the set of seen values
        - Return void 
    */
    set<int> values_seen;
    node_t * cur_node = head;
    node_t * prev_node = head;
    while (cur_node != nullptr){
        if (values_seen.count(cur_node->value) > 0) {
            // Value is already in the linked list
            // Remove this node from the linked list
            prev_node->next = cur_node->next;   
        }
        else {
            values_seen.insert(cur_node->value);
            prev_node = cur_node;
        }
        cur_node = (node_t *) cur_node->next;
    }
       
    return;
}

void print_linked_list(node_t * head){

    while(head != nullptr){
        if (head->next != nullptr)
            cout << head->value << "<-";
        else
            cout << head->value << endl;
        head = (node_t *) head->next;
    }

    return;
}


void test_remove_duplicates(){

    // Create a sample linked list:
    // 1->3->4->3->2->1
    // Expected result:
    // 1->3->4->2;
    node_t n1(1);
    node_t n2(2, &n1);
    node_t n3(3, &n2);
    node_t n4(4, &n3);
    node_t n5(3, &n4);
    node_t n6(1, &n5);
    node_t * linked_list = &n6;

    print_linked_list(linked_list);
    remove_duplicates(linked_list);
    print_linked_list(linked_list);

    return;
}

int main(){
    test_remove_duplicates();
    cout << "ALL TESTS PASSED" << endl;
    return 0;
}

File: remove_dups_from_linkedlist/remove_dups_from_linkedlist.py


class Node(object):

    def __init__(self, v, n=None):
        self.next = n
        self.value = v

def remove_duplicates(linked_list):

    cur_node = linked_list
    prev_node = cur_node
    visited = set()
    while cur_node is not None:
        if cur_node.value in visited:
            # If we have already seen the current
            # node's value, then skip it
            prev_node.next = cur_node.next
        else:
            # Otherwise, add it's value to the
            # set of seen values
            visited.add(cur_node.value)
            prev_node = cur_node
        cur_node = cur_node.next
    return

def test_remove_duplicates():
    # Very cursory test, not suitable for production

    n1 = Node(18)
    n2 = Node(2, n1)
    n3 = Node(3, n2)
    n4 = Node(2, n3)
    n5 = Node(3, n4)
    n6 = Node(5, n5)
    n6 = Node(1, n6)

    remove_duplicates(n6)

    print n6.next.next.next.next.value
    assert n6.next.next.next.next.value == 18


def main():
    test_remove_duplicates()
    print "ALL TESTS PASSED"

if __name__ == "__main__":
    main()
